Support narrow() on batch dim for NJT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143311
* __->__ #142063

Requested in https://github.com/pytorch/pytorch/issues/136270

cc @cpuhrsch @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

jbschlosser committed:
Support narrow() on batch dim for NJT

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"




cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"


Requested in https://github.com/pytorch/pytorch/issues/136270

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"


Requested in https://github.com/pytorch/pytorch/issues/136270

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser commented:
@soulitzer / @cpuhrsch this is ready now - the PR works in eager / compile without graph breaks for contiguous NJTs

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
+
+        # compute new sizes / strides from symbolic values
+        new_sizes = list(inp._values.size())
+        new_sizes[dim] = length
```
I guess this + using as_strided instead of narrow is to avoid pending unbacked symbols? Maybe worth adding a small comment :)

jbschlosser replied:
yep exactly, added a comment describing the situation. Essentially, we need all unbacked symints allocated to be directly accessible from the returned tensor's metadata (sizes, strides, storage offset) and not part of non-trivial expressions (e.g. as you'd get for the storage offset from calling narrow() with unbacked SymInt inputs)

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -1542,7 +1617,7 @@ def view_default(func, *args, **kwargs):
 
     # Ensure specified size still includes batch and ragged dims
-    if len(size) < 3 or not raggedness_matches(inp, size):
+    if len(size) < 2 or not raggedness_matches(inp, size):
```
Nice catch

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
+            end_val += inp._values.size(dim)
+        start_val = max(min(start_val, inp._values.size(dim)), 0)
+        end_val = max(min(end_val, inp._values.size(dim)), 0)
+        length_val = max(min(length_val, end_val - start_val), 0)
```
Unfortunate that we have to duplicate narrow input checking/manipulation logic here, but I guess maybe hard to avoid due to as_strided use.

jbschlosser replied:
Yeah agreed, I don't like this at all :( but indeed we aren't redispatching to narrow(), so we can't utilize the checks there

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
+    if operating_on_batch:
+        # batch dim narrowing requires custom logic involving offsets
+        out_kwargs = extract_kwargs(inp)
+        start_val, length_val = new_kwargs["start"], new_kwargs["length"]
```
Trying to think of weird edge cases (totally fine if normal narrow doesn't cover this either, but I guess we want parity?)

- What happens if length_val is negative? I didn't see any explicit check, and  [docs](https://pytorch.org/docs/stable/generated/torch.narrow.html ) say that it must "weakly positive" whatever that means 🤔.

- What happens if start_val is negative, but length_val > size of the tensor.

jbschlosser replied:
good call, will add test cases for these

>  it must "weakly positive" whatever that means

my understanding is that this allows for length=0? idk for sure though

soulitzer reviewed (approved):
Looks great! Some small comments

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8495,6 +8567,18 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
         op_match_fn=lambda device, op: (op.full_name in {"cdouble", "cfloat", "chalf"}),
         name="unimplemented_view_as_real",
     ),
+    # narrow(): unbacked SymInt bug with non-contig transposed inputs
```
Oof. Just curious - Is there an issue somewhere? If there is, maybe worth linking.

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
+            start_val += inp._values.size(dim)
+        if end_val < 0:
+            end_val += inp._values.size(dim)
+        start_val = max(min(start_val, inp._values.size(dim)), 0)
```
Hmm feels like it shouldn't be possible to be negative here (prior to max with 0)

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2164,6 +2164,30 @@ Tensor split_backward(
+  Tensor narrowed_grad_values = at::_nested_get_values(narrowed_grad);
+  TORCH_INTERNAL_ASSERT(
+      grad_values.dim() == narrowed_grad_values.dim(),
+      "Bug encountered in _nested_narrow_backward(); please open an issue");
```
nit: technically TORCH_INTERNAL_ASSERT already tells us to open an issue I think lol

jbschlosser replied:
ah okay, will remove then

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
+
+    if inp._lengths is not None or inp._ragged_idx != 1:
+        raise RuntimeError(
+            "narrow(): not yet supported for non-contiguous nested tensors on dim != 0"
```
maybe its my own fault, but I keep reading this as "tensors that are not contiguous on dim != 0" 
If it were instead worded as 
`not yet supported on dim != 0 for non-contiguous nested tensors` I would not make that mistake.

jbschlosser replied:
much clearer! fixed

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8386,6 +8449,14 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        op_match_fn=lambda device, op: (op.full_name == "select"),
+        sample_match_fn=lambda device, sample: ("batch_dim" in sample.name),
+        name="broken_select_backward_unbacked",
+    ),
```
This one is just code movement right.

jbschlosser replied:
yep from compile backward -> compile forward xfails

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -6022,6 +6022,67 @@ def test_narrow(self, device):
 
+    @skipIfTorchDynamo("Test compiles internally")
+    @skipCUDAIf(not SM70OrLater, "GPU capability is < SM70")
+    @torch._dynamo.utils.disable_cache_limit()
```
Woah do we always need to do this now? I'm not sure we do it for a lot of our other tests...

jbschlosser replied:
disable the cache limit? for a test with a bunch of OpInfo samples where compiler state resetting / recompiling happens a lot, I think it matters. the other tests should generally be fine though

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -6022,6 +6022,67 @@ def test_narrow(self, device):
+            self.assertEqual(out3_comp, nt_comp)
+
+        # length past the end
+        with self.assertRaisesRegex(RuntimeError, "exceeds dimension size"):
```
In theory this is the type of thing we could also handle via XFail with sample_match_fn right? 

Is it fair to say that we don't want to clutter that too much with basic input validity checks, and reserve those for actual bugs, features that are not implemented.

jbschlosser replied:
Yeah that's right - I have a mental TODO to use the error_inputs_func feature of OpInfos to handle this type of expected error checking failures. Ideally as you mentioned I'd want to avoid cluttering xfails with things that we'll never address and don't actually represent bugs

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -6022,6 +6022,67 @@ def test_narrow(self, device):
+    @torch._dynamo.utils.disable_cache_limit()
+    @dtypes(torch.float32)
+    @parametrize("env", ["eager", "compile", "compile_dynamic"])
+    def test_narrow_on_batch_dim(self, device, dtype, env):
```
I wonder how much overlap there is with this test and the narrow OpInfo one. 
Like could the "first few", "middle", and "last" batch items be formulated as sample inputs such that we can have two smaller tests that do more narrow things, e.g. `test_narrow_on_batch_dim_input_validation` and `test_narrow_on_batch_dim_narrow_of_narrow`.

jbschlosser replied:
> Like could the "first few", "middle", and "last" batch items be formulated as sample inputs?

Definitely and I should do that :) will fix

> And then we could reformalate this one to two smaller tests that do more specific things, e.g. test_narrow_on_batch_dim_input_validation and test_narrow_on_batch_dim_narrow_of_narrow.

yeah these are harder to test with OpInfo so I think it makes sense to break them out

jbschlosser replied:
oh I just realized the compile tests for narrow-on-narrow weren't actually compiling :p

changing them to actually compile fails them; investigating

jbschlosser replied:
okay i tracked this down to incorrect clamping. old logic was clamping (start, end) on inner values dim space, but we want to clamp on outer batch dim space. This fixed the data-dependent guard errors.

There's still some work to be done for non-contiguous NJTs apparently, but I've been deprioritizing those in general so I'll just land this as-is

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"


Requested in https://github.com/pytorch/pytorch/issues/136270

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

jbschlosser committed:
Update on "Support narrow() on batch dim for NJT"


Requested in https://github.com/pytorch/pytorch/issues/136270

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]
